docs/proposals/20210628-workloadspread.md

```diff
@@ -0,0 +1,223 @@
+
+Not like the workload type, WorkloadSpread can be used with those stateless workloads, such as CloneSet, Deployment and ReplicaSet.
+
+So users needn't learn the usage of a new workload and how to deploy and update their application by it.
+They only have to create WorkloadSpread CRs related to their existing workloads. Everything is just great!
+
+## Proposal
+
+WorkloadSpread definition looks like this:
+
+```yaml
+apiVersion: apps.kruise.io/v1alpha1
+kind: WorkloadSpread
+metadata:
+  namespace: ...
+  name: ...
+  # ...
+spec:
+  targetRef:
+    apiVersion: apps/v1 | apps.kruise.io/v1alpha1
+    kind: Deployment | CloneSet
+    name: workload-xxx
+  subsets:
+  - name: subset-a
+    requiredNodeSelectorTerm:
+      matchExpressions:
+      - key: topology.kubernetes.io/zone
+        operator: In
+        values:
+        - zone-a
+    preferredNodeSelectorTerms:
+    - weight: 50
+      preference:
+      matchExpressions:
+      - key: topology.kubernetes.io/zone
+        operator: In
+        values:
+        - zone-b
+    maxReplicas: 3
+    tolertions: []
+    patch:
+      metadata:
+        labels:
+          xxx-specific-label: xxx
+  - name: subset-b
+    requiredNodeSelectorTerm:
+      matchExpressions:
+      - key: topology.kubernetes.io/zone
+        operator: In
+        values:
+        - zone-b
+  scheduleStrategy:
+    type: Adaptive | Fixed
+    adaptive:
+      disableSimulationSchedule: false
+      rescheduleCriticalSeconds: 30
+status:
+  observedGeneration: 12
+  subsetStatuses:
+  - name: subset-a
+    missingReplicas: 2
+    creatingPods:
+      pod-xxx: "2021-04-27T08:34:11Z"
+    deletingPods:
+      pod-yyy: "2021-04-27T08:34:11Z"
+    subsetUnscheduledStatus:
+      unschedulable: true
+      unscheduledTime: 2021-04-27T08:34:11Z
+      failedCount: 1
+  - name: subset-b
+    missingReplicas: -1
+```
+
+**spec**:
+- `targetRef`: reference to a specific workload, it supports `Deployment` in `apps` group and `CloneSet` in `apps.kruise.io` group
+- `subsets`: list of subsets which defines different topologies
+  - `name`: name of this subset
+  - `requiredNodeSelectorTerm`: the additional node selector of this subset, it will be injected into each term of `pod.spec.affinity.nodeAffinity.requiredDuringSchedulingIgnoredDuringExecution.nodeSelectorTerms`
+  - `preferredNodeSelectorTerms`: the additional node selector of this subset, it will be injected into each term of `pod.spec.affinity.nodeAffinity.preferredDuringSchedulingIgnoredDuringExecution`
+  - `tolerations`: the additional tolerations of this subset, it will be injected into `pod.spec.tolerations`
+  - `patch`: the specified strategic patch body of this subset, it will be patched to Pods that belong to this subset
+  - `maxReplicas`: the maximum number of Pods in this subset, it can be an absolute number or percentage string. maxReplicas 
```

Comment:
I suggest to change `maxReplicas` to `replicas`. `maxReplica` means the controller will remove the extra Pods proactively which seems not to be the case for WorkloadSpread controller.  It should be a preferred number, not a max number.

Reply:
Controller will remove the extra Pods when the number of Pods belongs to this subset is less than maxReplicas, which is in line with our expectations.  Of course,  controller can't remove Pods directly rather than setting different deletion-cost annotation for Pods.
User can scale down subset's maxReplicas and workload's replicas to reduce the pressure on the subset or only release resources .